Resolve nest pub/sub subscriber token refresh issues

allenporter opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Create a full Credentials object (complete with refresh token, expiration time,
etc) so that the Pub/sub subscriber can handle token refreshes rather than
crashing after an hour.  While it is not ideal to have both Home Assistant and
Pub/sub handle token refresh, it significantly simplifies subscriber
troubleshooting to have it handle all credential management itself since it has no existing callback for getting
refreshed tokens.  Additionally, user reports that when credentials expire the subscriber may not notice, leading to
dropped "acks"s and duplicate delivery of events, since ACKs are best effort.

I previously attempted this change in #42947 and it did not work.  However, now it does
now so I must have not been creating the Credential correctly previously (e.g.
missing the expiry field in the parent class, which presumably causes the
client to know to refresh).  The workaround added in #43034 can likely be reverted after this change.

Updated google-nest-sdm version to v0.2.7
- See diffs: https://github.com/allenporter/python-google-nest-sdm/compare/v0.2.5...v0.2.7
- Motivation for this PR: Fixed a typo in the AbstractSubscriberFactory
- Context for other image related diffs: https://github.com/home-assistant/core/pull/44638 


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #44589
- This PR is related to issue: #42947
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Resolve pub/sub subscriber token refresh issues

Create a full Credentials object (complete with refresh token, expiration time,
etc) so that the Pub/sub subscriber can handle token refreshes rather than
crashing after an hour.  While it is not ideal to have both Home Assistant and
Pub/sub handle token refresh, it significantly simplifies subscriber
troubleshooting to have it handle all credential management itself since it has no existing callback for getting
refreshed tokens.  Additionally, user reports that when credentials expire the subscriber may not notice, leading to
dropped "acks"s and duplicate delivery of events, since ACKs are best effort.

I previously attempted this change, and it did not work.  However, now it does
now so I must have not been creating the Credential correctly previously (e.g.
missing the expiry field in the parent class, which presumably causes the
client to know to refresh).

allenporter commented:
Flaky test is due to broken build, not this PR, possibly fixed in https://github.com/home-assistant/core/pull/44687

MartinHjelmare commented on homeassistant/components/nest/api.py:
```diff
@@ -16,20 +20,37 @@ def __init__(
+        self._client_id = client_id
+        self._client_secret = client_secret
 
     async def async_get_access_token(self):
```
Is the library still using this method?

MartinHjelmare replied:
As long as Home Assistant can still update the token as needed when setting up the integration, it seems ok.

MartinHjelmare commented on tests/components/nest/test_api.py:
```diff
@@ -0,0 +1,106 @@
+        hass, DOMAIN, CLIENT_ID, CLIENT_SECRET, OAUTH2_AUTHORIZE, OAUTH2_TOKEN
+    )
+    session = config_entry_oauth2_flow.OAuth2Session(hass, config_entry, implementation)
+    return AsyncConfigEntryAuth(aioclient_mock, session, CLIENT_ID, CLIENT_SECRET)
```
We don't want to interact with integration details in tests. Just use the core interfaces to start the tests and patch in the periphery as needed.

MartinHjelmare replied:
We should not use `AsyncConfigEntryAuth` directly. We should just set up the integration and assert its interaction with the library by patching appropriately in the library.

allenporter replied:
Thanks, I interpreted your comment in the opposite direction.  This has been updated to exercise everything through setting up the integration.

(I can see how maintenance on common/core components would be much easier if we don't have to worry about updating tests with extra dependencies)

MartinHjelmare commented on tests/components/nest/test_api.py:
```diff
@@ -0,0 +1,106 @@
+    """Verify that the Credentials are created properly."""
+    auth = create_auth(hass, aioclient_mock, CONFIG_ENTRY_DATA)
+
+    creds = await auth.async_get_creds()
```
We want to patch the library so we can assert the return value of `async_get_creds` without patching `async_get_creds` and without calling `async_get_creds` directly.

allenporter replied:
For subscriber based tests.  Now using a `FakeSubscriberFactory` to capture the creds, which I think partially addresses this comment.  However, I'm not sure I agree this is the right test approach and there is an inconsistency with testing between the api vs subscriber.  Let me give some more context/background and we can see if there is an even better approach or future interface/interfaces I should improve since the AbstractAuth pattern may be failing.

The [AbstractAuth](https://github.com/allenporter/python-google-nest-sdm/blob/master/google_nest_sdm/auth.py) pattern is based on the recommended approach, but has a couple tweaks since it is used by two APIs:
- `request`: Public method to call the API
- `async_get_access_token`: Implemented by child class, used internally by `AbstractAuth`
- `async_get_creds: Used by the `GoogleNestSubscriber` when creating a subscriber.  At the moment this is considered part of the "public" API of `AbstractAuth`
- `AbstractSubscriberFactory` / `DefaultSubscriberFactory`: Used by `GoogleNestSubscriber` to actually create the subscriber, and it gets `creds` passed in

Question #1: For API based tests, I did not yet update these to change to a similar approach.   I'm taking your comment to mean that similarly there we shouldn't be testing `async_get_access_token` directly, but instead testing based on the `request` method?

Question #2: Do we need to improve the interface /design for subscriber auth? Currently `async_get_creds` seemed like the right level because its the "public" API for how the [GoogleNestSubscriber](https://github.com/allenporter/python-google-nest-sdm/blob/349619450e1fa4ac2de8ae0be282b0778e8365b0/google_nest_sdm/google_nest_subscriber.py#L132) gets the auth, differing from how the API based auth works.
- Maybe `AbstractAuth` is doing too much.  This may be case where it should prefer composition rather than inheritance (pull out auth creation into a standalone class not mixed in with request logic, breaking from the Home Assistant pattern).  This is keeping with he status quo of sharing auth logic between the API and susbcriber cases.
- Maybe the equivalent design would be to add an `async_get_creds` method to `AbstractSubscriberFactory` and callers implement that.  This would be a larger change to split out auth separately for API and Subscriber.
Do you have thoughts about this? Curious if you know what I mean about the public interfaces.

MartinHjelmare replied:
I'm not sure about the library design. I think you know that best. As long as we're using the Home Assistant API correctly and things work well I'm happy.

Regarding tests, I've tried to answer that above.

allenporter replied:
OK, all this was based on my total misunderstanding of your comment, thinking we were talking about "unit testing".

allenporter committed:
Clarify purpose of each Auth API method

allenporter committed:
Partially address PR feedback on testing

allenporter committed:
Test integration more black box style

- Exercise auth logic using the integration
- Share config entry setup with all other tests
- Revert manifest changes to bump library versions, since those fixes are not needed

allenporter committed:
Nest api test readability improvements

allenporter committed:
Fix patch import for nest tests

MartinHjelmare reviewed (approved):
Looks good!
